docs/Coding/mapgen.rst

```diff
@@ -0,0 +1,569 @@
+  | Ice    | Frozen    | \-        | Mountain | \-      | \-          | \-      | \-        |
+  +--------+-----------+-----------+----------+---------+-------------+---------+-----------+
+  | Tundra | Cold      | \-        | Mountain | \-      | \-          | \-      | \-        |
+  +--------+-----------+-----------+----------+---------+-------------+---------+-----------+
+  | Plains | Temperate | Green     | Mountain | \-      | \-          | \-      | \-        |
+  +--------+-----------+-----------+----------+---------+-------------+---------+-----------+
+
+Terrain patches expand outwards from a seed tile until the required tile
+properties are no longer met or a threshold in colatitude and height difference
+is reached (*Thr.* in the table). Ice, tundra, and plains/grassland are
+generated to fill in gaps and do not expand in patches.
+
+The algorithm to match the desired terrain properties to the ruleset-defined
+terrain types by first collecting all terrains with the required property. Then,
+types without at least some of the "preferred" property and types with a
+non-zero "avoided" property are removed from the set. Of the remaining terrains,
+one is picked at random, with a higher chance to be selected when the required
+property has a high value in the ruleset. If this search fails, it is resumed
+without the "preferred" property; if this fails again, the "avoided" property is
+also dropped.
+
+.. _mapgen-height-rivers:
+
+Rivers
+^^^^^^
+
+Rivers are generated by flowing them from springs chosen randomly on the map.
+Springs may not be frozen or low in the height map. They may not be ocean and
+there may not be a river in the area yet. The algorithm also tries to avoid
+springs in locations with many hills and mountains nearby, or in ice and
+deserts (according to the corresponding terrain properties: ``mountainous``,
+``frozen``, ``dry``).
+
+Once a spring is found, the river is flown from there one tile at a time.
+To decide which direction the river takes, the possible directions are tested in
+a series of tests until there is only one direction left. Some tests are fatal.
+This means that they can sort away all remaining directions. If they do so, the
+river is aborted. Here follows a description of the test series:
+
+Falling into itself: fatal
+  This is tested by keeping track of all tiles already used or evaluated while
+  creating the river. If a river comes close to one of these tiles, it is
+  falling back to itself and is aborted.
+
+Forming a 4-river-grid: fatal
+  A river may not form a grid with four rivers tiles next to each other.
+
+Highlands:
+  Rivers must not flow up in mountains or hills if there are alternatives.
+
+Adjacent ocean:
+  Rivers must flow down to coastal areas when possible.
+
+Adjacent river:
+  Rivers must flow down to areas near other rivers when possible:
+
+Adjacent highlands:
+  Rivers must not flow towards highlands (terrains with a non-zero
```

Comment:
s/terrains/terrain

Reply:
This is also about terrain "types" from the ruleset